File: Java/BOJ/괄호_9012.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;

public class °ýÈ£_9012 {
	public static void main(String args[]) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		int yesno = 0;
		
		for(int i=0; i<N; i++) {
			String s = br.readLine();
			for(int j=0; j<s.length(); j++) {
				if(s.substring(j, j+1).equals("(")) {
					yesno++;
				} else if(s.substring(j, j+1).equals(")")) {
					if(yesno==0) {
						yesno = Integer.MAX_VALUE;
						break;
					} else {
						yesno--;
					}
				}
			}
			if(yesno == 0) {
				System.out.println("YES");
			} else {
				System.out.println("NO");
			}
			yesno = 0;
		}
	}
}

File: Kotlin/BOJ/괄호_9012.kt
package com.example.kotlincodingtest.baekjoon

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.Stack
import java.util.StringTokenizer

fun main() = with(BufferedReader(InputStreamReader(System.`in`))){
    val bw = BufferedWriter(OutputStreamWriter(System.`out`))
    val stack = Stack<String>()
    val num: Int = readLine().toInt()
    for(n in 0 until num) {
        val st = StringTokenizer(readLine(), "")
        for(s in st.nextToken()) {
            stack.push(s.toString())
        }

        if(stack.size % 2 != 0) {
            bw.write("NO\n")
            stack.clear()
        } else {
            var countEnd = 0
            while(stack.size != 0) {
                if(stack.pop() == ")") {
                    countEnd++
                } else { //stack.pop == "("
                    if(countEnd == 0) {
                        stack.clear()
                    }
                    countEnd--
                }
            }
            bw.write(if(countEnd == 0) "YES\n" else "NO\n")
        }
    }
    bw.flush()
    bw.close()
    close()
}